invenio_jsonschemas/api.py

```diff
@@ -0,0 +1,164 @@
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Schema API.
+
+This collects JSON schemas of all modules. These should be located in the
+`jsonschemas` directory relative to the module root. It also generates schemas
+for all entries located under `jsonschemas/forms/*` by doing an `allOf` merge
+with the schema configured using `JSONSCHEMAS_BASE_SCHEMA`. The name of the
+results are formed by doing the `s|forms|records|` rename operation. No
+virtual schemas are generated in case of:
+
+- The target name already exists.
+- The base schema is not configured.
+
+.. note::
+    Be aware the the schema names do NOT include the `jsonschemas/` prefix.
+"""
+
+import copy
+import json
+import os
+import os.path
+import urlparse
+
+from flask import current_app, url_for
+
+from jsonpointer import resolve_pointer
+
+import jsonschema
+
+from speaklater import _LazyString
+
+from .registry import json_schemas as schema_files
+
+
+class InsecureSchemaLocation(Exception):
+
+    """The try to load a JSON schema from an insecure location."""
+
+
+def internal_schema_url(*parts):
+    """Return the URL to an internal JSON schema."""
+    return url_for(
+        'jsonschemas.schema',
+        path=os.path.join(*parts),
+        _external=True
+    )
+
+
+def get_schemas():
+    """Return a list of all available schemas.
+
+    This might include virtual/generated schemas.
+    """
+    schemas = set()
+    for f in schema_files.iterkeys():
+        schemas.add(f)
+
+        # also add a record for all forms
+        if f.startswith('forms' + os.path.sep):
+            f2 = f.replace('forms', 'records', 1)
+            schemas.add(f2)
+    return schemas
+
+
+def get_schema_data(f, default=None):
+    """Return schema data (dict) for a given schema.
+
+    In case the schema does not exist, `default` is returned.
+    """
+    # try to serve the resource
+    if f in schema_files:
+        with open(schema_files[f]) as resource:
+            return json.load(resource)
+
+    # now try to rewrite records to forms
+    base = current_app.config.get('JSONSCHEMAS_BASE_SCHEMA')
+    if base and f.startswith('records' + os.path.sep):
+        f2 = f.replace('records', 'forms', 1)
+        if f2 in schema_files:
+            return {
+                'allOf': [
+                    {'$ref': internal_schema_url(base)},
+                    {'$ref': internal_schema_url(f2)}
+                ]
+            }
+
+    # ok, not found, too bad
+    return default
+
+
+def get_schema_from_somewhere(uri):
```

Comment:
`somewhere` is not really nice name for API. Can you find a better name (e.g. `get_schema_from_uri`, `resolve_external_pointer`, ...)?